### DISCUSSION TITLE
Get weapons that player carried on death?

### DISCUSSION THREAD

**Discussion by thecatontheflat** [saul/demofile#253] [Questions] [answered]
Thank you for the effort of putting this library together!

Do you know if there is a way to get the items that the player carried when they were killed?

Below is the code that I tried

```
demoFile.gameEvents.on('player_death', e => {
    const victim = demoFile.entities.getByUserId(e.userid);
    console.log(victim.weapons);
});
```

This is always empty, but I guess it's because during at the `player_death` event a player technically does not have any weapons anymore.

I tried storing a "history" of a player by copying their weapons on `tickstart`, but it has a huge performance impact (getting all the weapons on each tick is expensive).

I was wondering if there was a way to access the data in a state at the beginning of the tick?

Thank you!

**Comment by saul** [comment]
Thanks for the kind words. 

You’re best off checking when the m_hMyWeapons table changes on any player (start listening after the entity is created), and keeping track of that rather than saving the state every tick.

Try that and let me know how you get on.

**Comment by thecatontheflat** [comment, marked as answer]
Thank you for pointing me in the right direction! 

It looks like I managed to achieve what I needed to with the following code (monitoring the entities on change). There is slight performance degradation, but I guess this is expected when you're listening to the entities changes (from 73 seconds to 90 seconds execution time).

```
  const tickPlayersWeapons = {
    // tick:userId: {
    //   old: [],
    //   new: [],
    // },
  };

  demoFile.entities.on('change', e => {
    if (e.tableName !== 'm_hMyWeapons') return;
    if (e.entity.constructor.name !== 'Player') return;

    const weaponOld = demoFile.entities.getByHandle(e.oldValue);
    const weaponNew = demoFile.entities.getByHandle(e.newValue);

    if (weaponOld) {
      tickPlayersWeapons[`${demoFile.currentTick}:${e.entity.userId}`] = tickPlayersWeapons[`${demoFile.currentTick}:${e.entity.userId}`] || { old: [], new: [] };
      tickPlayersWeapons[`${demoFile.currentTick}:${e.entity.userId}`].old.push(weaponOld.className);
    }

    if (weaponNew) {
      tickPlayersWeapons[`${demoFile.currentTick}:${e.entity.userId}`] = tickPlayersWeapons[`${demoFile.currentTick}:${e.entity.userId}`] || { old: [], new: [] };
      tickPlayersWeapons[`${demoFile.currentTick}:${e.entity.userId}`].new.push(weaponNew.className);
    }
  });

  demoFile.gameEvents.on('player_death', e => {
    const existingPlayer = playerStructure.getPlayer(players, demoFile, e.userid);
    const droppedWeapons = tickPlayersWeapons[`${demoFile.currentTick}:${e.userid}`];

    console.log(droppedWeapons.old);
  });
```

Cheers!
